src/topology/uniform_space/compare_reals.lean

```diff
@@ -0,0 +1,114 @@
+of uniform spaces but extending multiplication in an ad-hoc way.
+
+TODO:
+* Upgrade this isomorphism to a topological ring isomorphism.
+* Do the same comparison for p-adic numbers
+
+## Implementation notes
+
+The heavy work is done in `topology/uniform_space/abstract_completion` which provides an abstract
+caracterization of completions of uniform spaces, and isomorphisms between them. The only work left
+here is to prove the uniform space structure coming from the absolute value on ℚ (with values in ℚ,
+not referring to ℝ) coincides with the one coming from the metric space structure (which of course
+does use ℝ).
+
+## References
+
+* [N. Bourbaki, *Topologie générale*][bourbaki1966]
+
+## Tags
+
+real numbers, completion, uniform spaces
+-/
+
+open set function lattice filter cau_seq uniform_space
+
+/-- The metric space uniform structure on ℚ (which presupposes the existence
+of real numbers) agrees with the one coming directly from (abs : ℚ → ℚ). -/
+lemma rat.uniform_space_eq :
+  is_absolute_value.uniform_space (abs : ℚ → ℚ) = metric_space.to_uniform_space' :=
+begin
+  ext s,
+  erw [metric.mem_uniformity_dist, is_absolute_value.mem_uniformity],
+  split ; rintro ⟨ε, ε_pos, h⟩,
+  { use [ε, by exact_mod_cast ε_pos],
+    intros a b hab,
+    apply h,
+    rw [rat.dist_eq, abs_sub] at hab,
+    exact_mod_cast hab },
+  { obtain ⟨ε', h', h''⟩ : ∃ ε' : ℚ, 0 < ε' ∧ (ε' : ℝ) < ε, from exists_pos_rat_lt ε_pos,
+    use [ε', h'],
+    intros a b hab,
+    apply h,
+    rw [rat.dist_eq, abs_sub],
+    refine lt_trans _ h'',
+    exact_mod_cast hab }
+end
+
+/-- Cauchy reals packaged as a completion of ℚ using the absolute value route. -/
+noncomputable
+def rational_cau_seq_pkg : @abstract_completion ℚ $ is_absolute_value.uniform_space (abs : ℚ → ℚ) :=
+{ space := ℝ,
+  coe := (coe : ℚ → ℝ),
+  uniform_struct := by apply_instance,
+  complete :=  by apply_instance,
+  separation :=  by apply_instance,
+  uniform_inducing := by { rw rat.uniform_space_eq,
+                           exact uniform_embedding_of_rat.to_uniform_inducing },
+  dense := dense_embedding_of_rat.dense }
+
+/-- Type wrapper around ℚ to make sure the absolute value uniform space instance is picked up
+instead of the metric space one. We proved in rat.uniform_space_eq that they are equal,
+but they are not definitionaly equal, so it would confuse the type class system (and probably
+also human readers). -/
+def Q := ℚ
+
+instance : comm_ring Q := by unfold Q ; apply_instance
+instance : uniform_space Q := is_absolute_value.uniform_space (abs : ℚ → ℚ)
+
+/-- Real numbers constructed as in Bourbaki. -/
+def Bourbakiℝ : Type := completion Q
+
+instance : uniform_space Bourbakiℝ := completion.uniform_space Q
+
+/-- Bourbaki reals packaged as a completion of Q using the general theory. -/
+def Bourbaki_pkg : abstract_completion Q := completion.cpkg
+
+/-- The equivalence between Bourbaki and Cauchy reals-/
+noncomputable def compare_equiv : Bourbakiℝ ≃ ℝ :=
```

Comment:
This statement and the next ones are in the root namespace, as is the definition of `Q` above. Since these are quite general, I would put most of this file in a dedicated namespace to avoid clashes.

Reply:
This file is very unlikely to be imported anywhere, but I'll create a namespace anyway.